wagtail/core/telepath.py

```diff
@@ -16,21 +82,27 @@ def __init__(self):
         self.media = forms.Media(js=[
             versioned_static('wagtailadmin/js/telepath/telepath.js')
         ])
-        self.objects = {}
+
+        # Keep track of media declarations that have already added to self.media - ones that
+        # exactly match a previous one can be ignored, as they will not affect the result
+        self.media_fragments = set([str(self.media)])
+
+    def add_media(self, media):
+        media_str = str(media)
+        if media_str not in self.media_fragments:
+            self.media += media
+            self.media_fragments.add(media_str)
 
     def pack(self, obj):
         for cls in type(obj).__mro__:
             adapter = adapters.get(cls)
             if adapter:
-                break
-
-        if adapter is None:
-            raise Exception("don't know how to add object to JS context: %r" % obj)
+                return adapter.pack(obj, self)
 
-        self.media += adapter.get_media(obj, self)
-        return [adapter.js_constructor, *adapter.js_args(obj, self)]
+        # as fallback, try handling as an iterable
+        try:
+            return [self.pack(item) for item in obj]
+        except TypeError:
+            pass
 
-
-class Adapter(metaclass=MediaDefiningClass):
-    def get_media(self, obj, context):
-        return self.media
+        raise Exception("don't know how to pack object: %r" % obj)
```

Comment:
Could we use a more specific exception type? (probably a subclass of ``LookupError``?)

Reply:
Sure - json.dumps uses TypeError for the equivalent failure case, so I'd be inclined to do the same.

Reply:
...ah, but then the problem with TypeError is that it'll be caught and masked by the iterable handler (so it'll end up reporting a failure on the top-level list rather than the offending object) 🤔 

        # as fallback, try handling as an iterable
        try:
            return [self.pack(item) for item in obj]
        except TypeError:
            pass

        raise TypeError("don't know how to pack object: %r" % obj)

Reply:
Fixed in 7b62bfca2950f95e17c17bda09be06463616188c, although I might be overthinking this for the sake of keeping it as a TypeError - @kaedroho what do you think?

Reply:
Looks good to me!